src/txrequest.cpp

```diff
@@ -0,0 +1,238 @@
+    // then _BEST/REQUESTED, then _READY). Within the _READY entries, the best one (lowest priority) comes first.
+    // Thus, if an existing _BEST exists for the same txid that this entry may be preferred over, it must immediately
+    // precede the newly created _READY.
+    if (it == m_index.get<ByTxid>().begin() || std::prev(it)->m_txid != it->m_txid ||
+        std::prev(it)->GetState() == State::CANDIDATE_DELAYED) {
+        // This is the new best CANDIDATE_READY, and there is no IsSelected() entry for this txid already.
+        Modify<ByTxid>(it, [](Entry& entry){ entry.SetState(State::CANDIDATE_BEST); });
+    } else if (std::prev(it)->GetState() == State::CANDIDATE_BEST) {
+        uint64_t priority_old = std::prev(it)->ComputePriority(m_computer);
+        uint64_t priority_new = it->ComputePriority(m_computer);
+        if (priority_new < priority_old) {
+            // There is a CANDIDATE_BEST entry already, but this one is better.
+            Modify<ByTxid>(std::prev(it), [](Entry& entry){ entry.SetState(State::CANDIDATE_READY); });
+            Modify<ByTxid>(it, [](Entry& entry){ entry.SetState(State::CANDIDATE_BEST); });
+        }
+    }
+}
+
+void TxRequestTracker::ChangeAndReselect(typename TxRequestTracker::Index::index<ByTxid>::type::iterator it,
+    TxRequestTracker::State new_state)
+{
+    if (it->IsSelected()) {
+        auto it_next = std::next(it);
+        // The next best CANDIDATE_READY, if any, immediately follows the REQUESTED or CANDIDATE_BEST entry in the
+        // ByTxid index.
+        if (it_next != m_index.get<ByTxid>().end() && it_next->m_txid == it->m_txid &&
+            it_next->GetState() == State::CANDIDATE_READY) {
+            // If one such CANDIDATE_READY exists (for this txid), convert it to CANDIDATE_BEST.
+            Modify<ByTxid>(it_next, [](Entry& entry){ entry.SetState(State::CANDIDATE_BEST); });
+        }
+    }
+    Modify<ByTxid>(it, [new_state](Entry& entry){ entry.SetState(new_state); });
+    assert(!it->IsSelected());
+}
+
+bool TxRequestTracker::MakeCompleted(typename TxRequestTracker::Index::index<ByTxid>::type::iterator it)
+{
+    // Nothing to be done if it's already COMPLETED.
+    if (it->GetState() == State::COMPLETED) return true;
+
+    if ((it == m_index.get<ByTxid>().begin() || std::prev(it)->m_txid != it->m_txid) &&
+        (std::next(it) == m_index.get<ByTxid>().end() || std::next(it)->m_txid != it->m_txid ||
+        std::next(it)->GetState() == State::COMPLETED)) {
+        // This is the first entry for this txid, and the last non-COMPLETED one. There are only COMPLETED ones left.
+        // Delete them all.
+        uint256 txid = it->m_txid;
+        do {
+            it = Erase<ByTxid>(it);
+        } while (it != m_index.get<ByTxid>().end() && it->m_txid == txid);
+        return false;
+    }
+
+    // Mark the entry COMPLETED, and select the best best entry if needed.
```

Comment:
259fb92

nit: The first-sorted _READY ?

Reply:
Done.